### ISSUE TITLE
Syntactic rule to ban non-tuple val patterns

### ISSUE THREAD

**Issue by fommil** [scalacenter/scalafix#716]
```scala
val Right(noproblemo) = ... thing that returns `Either`...
```

is a bang at runtime waiting to happen. Should be banned.

Tuples should be special cased and allowed. The general case that should be allowed is a **total** pattern match, but there is no way to know that in scala.

**Comment by olafurpg** [comment]
This is a great idea! I think this can be part of the `DisableSyntax` rule under option `DisableSyntax.noValPatterns: Boolean`.

The configuration option can be added here 
https://github.com/scalacenter/scalafix/blob/6d48e0c419a63956274cfcc8e9bb23601c75ccfa/scalafix-core/shared/src/main/scala/scalafix/internal/config/DisableSyntaxConfig.scala#L18

and the check can be added somewhere around here
https://github.com/scalacenter/scalafix/blob/6d48e0c419a63956274cfcc8e9bb23601c75ccfa/scalafix-core/shared/src/main/scala/scalafix/internal/rule/DisableSyntax.scala#L57

A good place to test this would be here https://github.com/scalacenter/scalafix/blob/6d48e0c419a63956274cfcc8e9bb23601c75ccfa/scalafix-tests/input/src/main/scala/test/DisableSyntax.scala

**Comment by mcanlas** [comment]
What do you mean by "total pattern match"?

**Comment by fommil** [comment]
total in the sense of a total function, as opposed to a partial function.

Actually we can do better than just tuples in the "allowed" list... we can also allow synthetic case class unapply since they are also total. We may be able to detect this by seeing if the `unapply` returns a `Some` (as opposed to some other structure like `Option` or a structural thing).

**Comment by olafurpg** [comment]
If you want more sophisticated type analysis then this will require more expertise. I think a syntactic rule that prohibits any non-tuple pattern in vals would be useful in itself

**Comment by olafurpg** [comment]
I updated the PR to be about adding a syntactic rule. The more advanced case which works on types boils down to implementing exhaustivity checking, which is a research topic in itself and is already implemented in the compiler. I'm sure you can get those warnings in the compiler by writing a compiler plugin like https://github.com/oleg-py/better-monadic-for that removed the synthetic `@unchecked` annotation for val desctructuring

```scala
➜  target git:(v1) cat foo.scala
object a {
  val List(1) = List(3)
}
➜  target git:(v1) scalac -Xprint:typer foo.scala
[[syntax trees at end of                     typer]] // foo.scala
package <empty> {
  object a extends scala.AnyRef {
    def <init>(): a.type = {
      a.super.<init>();
      ()
    };
    <synthetic> <artifact> private[this] val x$1: Unit = 
(scala.collection.immutable.List.apply[Int](3): List[Int]
 @unchecked) match { // <-- synthetic @unchecked
      case scala.collection.immutable.List.unapplySeq[Int](<unapply-selector>) <unapply> (1) => ()
    }
  }
}
```

**Comment by perok** [comment]
I will take a look at this.

**Comment by joan38** [comment]
Hi everyone,
This sounds like a nice rule but detecting the total matching sounds quiet high to do.
@perok do you have any idea how you would solve this issue?

**Comment by fommil** [comment]
@joan38 we ended up just deciding that the rule should only allow tuples and that everything else is suspicious

**Comment by olafurpg** [comment]
@joan38 the best solution to reject non-total pattern matches would be to change the compiler to not emit `@unchecked` for val deconstruction
```scala
cat total.scala
object a {
  val List(a, b) = List(1)
}
➜  target git:(master) scalac -Xprint:parser total.scala
[[syntax trees at end of                    parser]] // total.scala
package <empty> {
  object a extends scala.AnyRef {
    def <init>() = {
      super.<init>();
      ()
    };
    private[this] val x$1 = List(1): @scala.unchecked match { // <-- unchecked
      case List((a @ _), (b @ _)) => scala.Tuple2(a, b)
    };
    val a = x$1._1;
    val b = x$1._2
  }
}
```

**Comment by olafurpg** [comment]
Fixed in #725
